include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -254,17 +350,32 @@ __parallel_for(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&&
                _Ranges&&... __rngs)
 {
     using _CustomName = oneapi::dpl::__internal::__policy_kernel_name<_ExecutionPolicy>;
-    using _ForKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<_CustomName>;
-
-    return __parallel_for_submitter<_ForKernel>()(::std::forward<_ExecutionPolicy>(__exec), __brick, __count,
-                                                  ::std::forward<_Ranges>(__rngs)...);
+    using _ForKernelSmall =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__parallel_for_small_kernel<_CustomName>>;
+    using _ForKernelLarge =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__parallel_for_large_kernel<_CustomName>>;
+
+    using __small_submitter = __parallel_for_small_submitter<_ForKernelSmall>;
+    using __large_submitter = __parallel_for_large_submitter<_ForKernelLarge, _Ranges...>;
+    // Compile two kernels: one for small-to-medium inputs and a second for large. This avoids runtime checks within a single
+    // kernel that worsen performance for small cases.
+    if (__count < __large_submitter::__estimate_best_start_size(__exec))
```

Comment:
I am not sure that `__large_submitter` able to calculate to best size in all cases and it doesn't depends on Kernel code and it's logic at all. So may be we should have ability to customize this condition check to?

Reply:
Its an interesting point.  I'm trying to think about how different workloads / algorithms would impact this number...

I think the decision is largely dependent on overhead vs memory bandwidth optimization.  Its possible that more computation would make this less important because we are less reliant on memory bandwidth.  However, depending on user provided call-ables in the library makes this very difficult to make good decisions, unless there are some APIs which we know are always very high computationally that use parallel_for internally (I don't know of any).  

Another aspect to consider, the larger the size of the minimum type size in the input ranges, the fewer iterations would be run by the large submitter.  At the limit, I imagine there is no advantage of the large submitter when we only would run a single iteration.  This is knowable at compile time, perhaps we should detect this case and always choose the small submitter when a single iteration would be used.

Reply:
One thing to note is that the performance difference at this "estimated point" is very small between the two versions so computing an optimal point to switch for each algorithm does not bring us any significant benefit and would likely be overtuned for a specific case / architecture, so I am in favor of leaving it as it currently is. The improvements are really observed once we scale to large sizes.

With regards to @danhoeflinger's comment, I agree that a user providing a heavily compute intensive operator might minimize the observed benefit if we become compute bound. I do not think we have encountered such a case yet, but I do not think there is a high risk of performance loss in this scenario although there may be small differences around this estimated point.

Good point on the case where we load one element per item. I will look into adding this as a compile time decision.

Reply:
This makes sense to me.  Thanks for the reply.

Reply:
I have added a compile time decision to only compile the basic kernel if the iters per work item is 1 for the large kernel. I benchmarked with a 128-bit type and confirmed performance is the same.